Prevent potential GPU OOM in R2C with split retry

thirtiseven opened the pull request:
Fixes https://github.com/NVIDIA/spark-rapids/issues/14018

### Description

https://github.com/NVIDIA/spark-rapids/pull/13842 added retry for R2C `convert` to prevent Host OOM. Following on from that, this pr aimed to add split and retry for GPU OOM when copying the converted results to GPU. 

This PR:
- Introduced `HostColumnarBatchWithRowRange`, a wrapper for host columns that supports logical slicing without copying underlying host memory. This allows splitting a large host batch into smaller chunks for transfer.
- Updated `GpuRowToColumnarExec` to use split and retry when a GPU OOM occurs during transfer.

Note that when we split a batch into two halves, we can't free them until both halves are processed. So, the first half just borrows the data, but we pass the ownership of the host columns to the second half. This ensures the host memory stays alive exactly as long as needed and is freed when the last split is closed.

**Performance tests:**

Benchmark measures end-to-end time for R2C conversion + GPU aggregation on 100M rows. Data is created via sc.parallelize + createDataFrame to force CPU rows through GpuRowToColumnarExec. Three test groups:

  - String columns — (Int, String, Int), forces non-codegen RowToColumnarIterator path (PR's code path)
  - Nested columns — (Int, Array[Int], Struct[Int, String]), same path
  - Fixed-width columns — (Int, Int, Int, Double), codegen path (control, unaffected by this PR)

  Each group: 5 warmup + 10 timed runs, reporting median. spark-shell --master local[4] --driver-memory 8g.

| Configuration | String cols | Nested cols | Fixed-width (control) |
  |---------------|-------------|-------------|-----------------------|
  | main (baseline) | 8,416 ms | 19,349 ms | 4,025 ms |
  | This PR | 8,611 ms | 20,040 ms | 4,044 ms |
  | Overhead | +2.3% | +3.6% | +0.5% (noise) |

benchmark code: 
[r2c_split_retry_benchmark.scala.zip](https://github.com/user-attachments/files/26621101/r2c_split_retry_benchmark.scala.zip)
   
### Checklists

Documentation
- [ ] Updated for new or modified user-facing features or behaviors
- [x] No user-facing change

Testing
- [x] Added or modified tests to cover new code paths
- [ ] Covered by existing tests
      (Please provide the names of the existing tests in the PR description.)
- [ ] Not required

Performance
- [x] Tests ran and results are added in the PR description
- [ ] Issue filed with a link in the PR description
- [ ] Not required

thirtiseven commented:
@greptile full review

thirtiseven commented:
build

thirtiseven commented on sql-plugin/src/main/scala/com/nvidia/spark/rapids/HostColumnarBatchWithRowRange.scala:
```diff
@@ -0,0 +1,335 @@
+          dataSlice.setByte(0L, 0.toByte)
+        } else {
+          dataSlice = origData.slice(startByte.toLong, dataLen.toLong)
+        }
```
good catch

firestarman commented on sql-plugin/src/main/java/com/nvidia/spark/rapids/GpuColumnVector.java:
```diff
@@ -356,6 +356,19 @@ public ColumnarBatch tryBuild(int rows) {
+        hostColumns = buildHostColumns();
+      }
+      HostColumnVector[] ret = hostColumns;
+      hostColumns = null;
```
Setting `hostColumns` to null will lead to trigger the call to "buildHostColumns" again when calling this "buildHostColumnsWithoutOwnership" API again, then an expcetion will be thrown out.

firestarman commented on sql-plugin/src/main/scala/com/nvidia/spark/rapids/HostColumnarBatchWithRowRange.scala:
```diff
@@ -0,0 +1,333 @@
+    val startRow: Int,
+    val numRows: Int,
+    val dataTypes: Array[DataType],
+    private var ownsHostColumns: Boolean) extends AutoCloseable {
```
This is a little confused. I would suggest leveraging `refCount` instead of this boolean for buffers lifecycle management.

thirtiseven commented on sql-plugin/src/main/java/com/nvidia/spark/rapids/GpuColumnVector.java:
```diff
@@ -356,6 +356,19 @@ public ColumnarBatch tryBuild(int rows) {
+     * Build host columns and transfer ownership to the caller. The builder will not close
+     * the returned host columns.
+     */
+    public HostColumnVector[] buildHostColumnsWithoutOwnership() {
```
Yes this can be removed if we use the refCount approach.

firestarman commented:
@greptile can you review it again ?

firestarman commented:
@greptileai can you review it again ?

firestarman commented on tests/src/test/scala/com/nvidia/spark/rapids/RowToColumnarIteratorRetrySuite.scala:
```diff
@@ -60,4 +80,84 @@ class RowToColumnarIteratorRetrySuite extends RmmSparkRetrySuiteBase {
+    assertThrows[GpuSplitAndRetryOOM] {
+      row2ColIter.next()
+    }
+  }
```
We need tests for nested types

firestarman commented on sql-plugin/src/main/scala/com/nvidia/spark/rapids/HostColumnarBatchWithRowRange.scala:
```diff
@@ -0,0 +1,338 @@
+   * the columns and closing it will not free them.
+   */
+  override def close(): Unit = {
+    hostColumns.foreach { hcv =>
```
Use `safeClose` instead

thirtiseven replied:
done.

firestarman commented on sql-plugin/src/main/scala/com/nvidia/spark/rapids/HostColumnarBatchWithRowRange.scala:
```diff
@@ -0,0 +1,338 @@
+
+  /**
+   * Create a HostColumnarBatchWithRowRange.
+   * The host columns will be closed when this instance is closed.
```
Can you add comments about how "startRow" is specified for this constructor ?

thirtiseven replied:
Added some comments

firestarman commented on sql-plugin/src/main/scala/com/nvidia/spark/rapids/HostColumnarBatchWithRowRange.scala:
```diff
@@ -0,0 +1,338 @@
+      dataTypes: Array[DataType]): HostColumnarBatchWithRowRange = {
+    require(numRows >= 0, s"numRows must be non-negative, got $numRows")
+    if (hostColumns.nonEmpty) {
+      require(hostColumns(0).getRowCount.toInt == numRows,
```
Shouldn't we check all the columns ?

thirtiseven replied:
Updated.

thirtiseven commented on sql-plugin/src/main/scala/com/nvidia/spark/rapids/HostColumnarBatchWithRowRange.scala:
```diff
@@ -0,0 +1,338 @@
+ * @param startRow The starting row index (inclusive) for this range
+ * @param numRows The number of rows in this range
+ * @param dataTypes The Spark data types for each column
+ * @param ownsHostColumns Whether this instance owns (should close) the host columns
```
Thanks, removed.

thirtiseven commented on sql-plugin/src/main/scala/com/nvidia/spark/rapids/HostColumnarBatchWithRowRange.scala:
```diff
@@ -0,0 +1,338 @@
+      hostColumns: Array[HostColumnVector],
+      numRows: Int,
+      dataTypes: Array[DataType]): HostColumnarBatchWithRowRange = {
+    require(numRows >= 0, s"numRows must be non-negative, got $numRows")
```
Removed.

firestarman commented on sql-plugin/src/main/scala/com/nvidia/spark/rapids/HostColumnarBatchWithRowRange.scala:
```diff
@@ -0,0 +1,339 @@
+   * It sets startRow to 0 since the host columns contain exactly the rows we need.
+   * For split batches with non-zero startRow, use the private constructor via splitInHalf.
+   *
+   * The host columns will be closed when this instance is closed.
```
This seems not true, since a HostColumnarBatchWithRowRange will have the refCoung increased by 1. The input hostColumns still have to be closed separately even  this instance is closed.

thirtiseven commented on tests/src/test/scala/com/nvidia/spark/rapids/RowToColumnarIteratorRetrySuite.scala:
```diff
@@ -60,4 +83,200 @@ class RowToColumnarIteratorRetrySuite extends RmmSparkRetrySuiteBase {
+      // String column
+      val str = UTF8String.fromString(s"test_string_$i")
+      // Array column
+      val arr = ArrayData.toArrayData(Array(i, i + 1, null, i + 2))
```
done.

firestarman commented:
Overall looks good to me. But nice to have more eyes on this.

abellina commented on sql-plugin/src/main/scala/com/nvidia/spark/rapids/HostColumnarBatchWithRowRange.scala:
```diff
@@ -0,0 +1,340 @@
+ * @param numRows The number of rows in this range
+ * @param dataTypes The Spark data types for each column
+ */
+class HostColumnarBatchWithRowRange private (
```
Should we add the slice functionality to cuDF instead of spark-rapids?

abellina replied:
@revans2 any preference?

abellina commented on sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuRowToColumnarExec.scala:
```diff
@@ -665,29 +684,57 @@ class RowToColumnarIterator(
+          numInputRows += rowCount
+          ret
+        } else {
+          val it = RmmRapidsRetryIterator.withRetry(
```
nit, import `RmmRapidsRetryIterator.withRetry`, like it was done for `withRetryNoSplit`

thirtiseven replied:
Done.

abellina commented on sql-plugin/src/main/scala/com/nvidia/spark/rapids/HostColumnarBatchWithRowRange.scala:
```diff
@@ -0,0 +1,340 @@
+ *
+ * Memory management uses reference counting: each instance increments the reference count
+ * of the host columns on construction and decrements it on close. The host columns are
+ * freed when the last reference is closed.
```
we should add that validity and offset buffers are copied, not logically sliced, in this description.

thirtiseven commented:
build

thirtiseven commented:
Hi @abellina could you take another look?

abellina commented on sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuRowToColumnarExec.scala:
```diff
@@ -651,29 +685,57 @@ class RowToColumnarIterator(
+          // Return the first batch now and keep the iterator for subsequent output batches.
+          // This ensures we only transfer one split at a time (avoid multiple device allocations).
+          closeOnExcept(it.next()) { first =>
+            pendingBatchIter = if (it.hasNext) it else Iterator.empty
```
nit, could just set pendingBatchIter = it, without checking, since we should be gating all the other logic in that it.hasNext returns false

thirtiseven commented on sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuRowToColumnarExec.scala:
```diff
@@ -651,29 +685,57 @@ class RowToColumnarIterator(
+          HostColumnarBatchWithRowRange(hostColumns, rowCount, dataTypes)
         }
 
+        if (localGoal.isInstanceOf[RequireSingleBatchLike]) {
```
done.

thirtiseven committed:
Split retry for R2C GPU OOM during host-to-GPU transfer

Rebased on top of #14428's per-batch retry mechanism. Adds GPU OOM
split retry for the host-to-GPU transfer step using
HostColumnarBatchWithRowRange, which supports logical slicing of
host columns without copying underlying memory.

Signed-off-by: Haoyang Li <haoyangl@nvidia.com>

thirtiseven committed:
Update stale comment about tryBuild retry

Signed-off-by: Haoyang Li <haoyangl@nvidia.com>

thirtiseven committed:
address comments

Signed-off-by: Haoyang Li <haoyangl@nvidia.com>

thirtiseven committed:
remove benchmark code

Signed-off-by: Haoyang Li <haoyangl@nvidia.com>

thirtiseven commented on sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuRowToColumnarExec.scala:
```diff
@@ -728,29 +751,56 @@ class RowToColumnarIterator(
+              pendingBatchIter = it
+              numInputRows += rowCount
+              first
+            }
```
Thanks for flagging this, but I think the leak is already prevented by the retry framework itself.

`pendingBatchIter` is a `RmmRapidsRetryAutoCloseableIterator` whose underlying `AutoCloseableAttemptSpliterator` registers an `onTaskCompletion` listener at construction time (`RmmRapidsRetryIterator.scala` ~L499-L504) that calls `closeInternal()` to `safeClose` the entire `attemptStack`. The split halves produced by `splitHostBatchInHalf` are pushed onto that same `attemptStack` (L571), so on task cancellation any unconsumed `HostColumnarBatchWithRowRange` instances are closed by the existing listener.

Adding another `onTaskCompletion` hook in `RowToColumnarIterator` would just duplicate that cleanup path, so I'll leave this as is.

thirtiseven commented:
Closing for priority
